### PR TITLE
API - Pagination chapter

### PR DESCRIPTION
**Pull request by lovro-bikic**
Task: [#356](https://app.productive.io/1-infinum/tasks/1670946)

#### Aim

The general aim is to rewrite the existing API guide and split it into multiple chapters. One of the chapters should explain pagination.

#### Solution

Add a pagination chapter. Explain what is pagination, why we use it, how to enable pagination, and different pagination types.

I'm requesting a review only from TLs and LEs for now to get initial feedback. I'll mention the rest of the team after the review so they can provide feedback too if they want to.

### CODE REVIEW

File: API/Pagination.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+When page number param is omitted, you should respond with the first page. Respond with an error if the param is lesser than 1, but if it's greater than the total number of pages then return the last page. Returning the last page instead of responding with an error prevents an edge case when pages disappear due to records being deleted but clients still requesting the missing page (which they think exists).
+
+Response should optionally include pagination metadata:
+- current page (in case it differs from the requested page)
+- total number of pages
+- total number of items
+This information is optional because it won't always be used by clients, and, since it typically requires executing an additional SQL COUNT query, it can be a costly operation.
+
+Since this type of pagination allows clients to request any page, it is usually reperesented in the UI with a paginator where you can select a page (e.g. `<- 2 3 4 5 6 ->`). If this is the kind of UI you're implementing pagination for, then page-based type is the right choice.
```
Comment (lovro-bikic) [inline API/Pagination.md:33]: I think it'd be useful to add pictures of UI elements appropriate for each pagination type, what do you think?

**Comment by lovro-bikic** [comment]
I didn't mention any gems, should I mention [pagy](https://github.com/ddnexus/pagy) since that is what we currently use for page-based pagination?

**Comment by melcha** [comment]
Shouldn't we also list gems we use for pagination? Pagy, kaminari... Not sure what you guys used for cursor pagination.

File: API/Pagination.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+Some API endpoints are designed to return a list of resources. We call those *list* or *index* endpoints.
+
+Servers can in practice have thousands of resources and in theory return all of them when asked to, but this can be costly, slow and impractical. It's costly and slow for the server to fetch everything from the database and serialize it, and for the client to download it and parse the response. It is also impractical because in most cases users don't care about all resources and instead only want to see top results (e.g. if the endpoint is used for search).
+
+Rather than showing everything at once, we can limit the number of resources returned in a request and provide a way to fetch the rest. In other words, we can page through resources.
+
+There are different ways to enable pagination and to paginate resources, however the following paragraphs are limited to our most commonly used methods. That is not to say these are the only methods available, you are encouraged to use something else if it benefits your project.
```
[the remark this answers is not in the document]
Comment (cilim) [reply API/Pagination.md:7]: then we'll have to take care of the resource health, whether its even alive or not 😕

File: API/Pagination.md (hunk 1 above)
Comment (cilim) [reply API/Pagination.md:33]: what would be the image for infinite scroll pagination? 😄

**Comment by cilim** [comment]
> Maybe we should mention a situation we had on Loop where we had 'missing' elements because the records weren't always sorted by some unique attribute (id).
> 
> If we had 5 records sorted by e.g. count column:
> A - 5
> B - 4
> C - 4
> D - 3
> E - 1
> 
> and you had page size 2, we had this issue:
> 
> page 1 - A B
> page 2 - B D
> 
> and record C was not shown because it was the "3rd record in total" on 1st page and "2nd record in total" on the second page.

@nikajukic would that mean that the fix is to always sort by `_whateverYouNeed_`,`_some_unique_column` ?

File: API/Pagination.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+In most cases, pagination is enabled through query string. Query string lets clients pass additional parameters which are not appropriate to send in headers or the body. A URL with pagination applied might look like this: `/api/v1/countries?page=5`. In the example, we're fetching countries and requesting the 5th page of results. Query string is the preferred way to enable pagination and also the one we use.
+
+## Types
+
+There are many pagination types, and working with APIs you'll find different implementations. We'll stick to two types which are the most prevalent.
+
+### Page-based
+
+Page-based pagination divides a set of records into pages of equal size (except for the last page which can have less records than other pages). A number is assigned to every page, starting with 1.
```
[the remark this answers is not in the document]
Comment (cilim) [reply API/Pagination.md:19]: added to L#41

File: API/Pagination.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+API should support client-set page size and page number params. An endpoint with pagination params applied could look like this: `/api/v1/countries?page[size]=50&page[number]=3`. In the example, the client requests the 3rd page of countries with maximum 50 records.
```
[the remark this answers is not in the document]
Comment (cilim) [reply API/Pagination.md:21]: added

**Comment by nikajukic** [comment]
@cilim yes, we're always sorting by ID additionally.

**Comment by cilim** [comment]
phew..I think I caught most of the comments and discussed some in the office. Please go over the PR once again and let me know if it's ok now or should we change something before sending for proofreading to Eva 🙂 

P.S. I think it would be nice to keep these chapters as agnostic as possible, and mention gems only in the chapter for relevant gems.

File: API/Pagination.md [hunk 5]
```diff
@@ -8,7 +8,7 @@ There are different ways to enable pagination and to paginate resources, however
 
 ## Enabling pagination
 
-In most cases, pagination is enabled through query string. Query string lets clients pass additional parameters which are not appropriate to send in headers or the body. A URL with pagination applied might look like this: `/api/v1/countries?page=5`. In the example, we're fetching countries and requesting the 5th page of results. Query string is the preferred way to enable pagination and also the one we use.
+In most cases, pagination is enabled through query string. Query string lets clients pass additional parameters which are not appropriate to send in headers or the body. A URL with pagination applied might look like this: `/api/v1/countries?page=5`. In the example, we're fetching countries and requesting the 5th page of results. Query string is the preferred way to enable pagination, but you can keep it also enabled by default, where a request without pagination query strings is handled as `page=1`.
```
Comment (melcha) [inline API/Pagination.md:11]: I'd say `you should keep it also enabled by default`, you can sounds like it's nice to have.

**Comment by melcha** [comment]
> P.S. I think it would be nice to keep these chapters as agnostic as possible, and mention gems only in the chapter for relevant gems.

What do you mean by chapter for relevant gems?

File: API/Pagination.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (d4be4st) [reply API/Pagination.md:19]: you are not using two different keywords on different lines for the same thing. It might be a bit confusing for some.

**Review by d4be4st — CHANGES_REQUESTED** [review]

**Pull request merged**